rfcs/text/0032-structured-benchmark-programs.md

```diff
@@ -0,0 +1,224 @@
+- *Iterative Quantum Phase Estimation (iQPE)*: Similar to iQFT, iQPE allocates a single qubit (in addition to the phase qubit "psi") and reuses it across iterations to estimate the phase of a unitary operator.
+- *Shor's Algorithm*: The primitives used in Shor's algorithm can similarly be implemented using dynamic qubit reallocation to minimize the total number of qubits required.
+- *Toffoli-heavy Circuits*: Toffoli gates in quantum circuits can be implemented more efficiently using ancilla qubits. Circuits using a larger number of Toffoli gates, such as many arithmetic circuits, can benefit from dynamic qubit allocation to allocate and deallocate ancilla qubits repeatedly inside loop bodies.
+
+## Loosely Coupled Hybrid Programs
+
+*Variational and hybrid algorithms naturally combine quantum circuits with classical optimization loops. They are loosely coupled because the quantum and classical parts interact only at well-defined synchronization points.*
+
+- *VQE*: A variational quantum eigensolver algorithm that uses a classical optimizer to minimize the energy of a parameterized quantum circuit which implements several iterations of some ansatz. ([Peruzzo et al., 2014](https://arxiv.org/abs/1304.3061))
+- *QAOA*: A quantum approximate optimization algorithm that uses a classical optimizer to find optimal parameters for a parameterized quantum circuit. It iteratively applies a "problem" and "mixer" operator using loop structures. ([Farhi et al., 2014](https://arxiv.org/abs/1411.4028))
+
+## Fault-Tolerant State Preparation and Error Correction
+
+*Quantum Error Correction (QEC) is one of the most important applications of structured control flow in quantum computing. These benchmark programs implement QEC protocols that involve structured operations at any point in the program.*
+
+- *Magic State Distillation*: Magic state distillation protocols utilize loops and conditionals to iteratively improve the fidelity of magic states, which are essential for fault-tolerant quantum computing. ([Bravyi & Kitaev, 2005](https://arxiv.org/abs/quant-ph/0403025))
+- *Logical Sate Preparation*: Preparing logical qubits in specific states  using QEC codes involves structured operations to ensure fault tolerance.
+- *Syndrome Measurement and Correction*: QEC protocols involve measuring error syndromes and applying corrections based on the measurement outcomes, utilizing conditionals to determine the appropriate corrective actions.
+
+---
+
+Each of the benchmark programs listed above should be implemented individually to achieve a comprehensive suite that covers a wide range of structured control flow scenarios.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are several potential drawbacks to consider with this proposal:
+
+- Maintenance Overhead: The addition of a new benchmark suite requires ongoing maintenance to ensure that the benchmarks remain relevant and up-to-date with the latest advancements in quantum computing and compiler technologies.
```

Comment:
We can also spin up a separate jeff-bench repo or similar, at least to avoid any challenges managing dependencies for the core jeff code from the benchmark generation code.

Reply:
I think we should be good here. We can use inline script metadata for the Python scripts that generate benchmarks. `uv` has great support for these.
See https://docs.astral.sh/uv/guides/scripts/#declaring-script-dependencies for some good documentation on that.